dotCMS/src/main/java/com/dotcms/publisher/util/dependencies/ConcurrentDependencyProcessor.java

```diff
@@ -0,0 +1,200 @@
+    @Override
+    public synchronized void addAsset(final Object asset, final PusheableAsset pusheableAsset) {
+
+        final boolean added = addRequestToProcess(asset, pusheableAsset);
+
+        if (added) {
+            queue.add(new DependencyProcessorItem(asset, pusheableAsset));
+        }
+    }
+
+    private <T> boolean addRequestToProcess(final T asset, final PusheableAsset pusheableAsset) {
+        final String assetKey = DependencyManager.getKey(asset);
+        Set<String> set = assetsRequestToProcess.get(pusheableAsset);
+
+        if (set == null) {
+            set = new ConcurrentHashSet<>();
+            assetsRequestToProcess.put(pusheableAsset, set);
+        }
+
+        return set.add(assetKey);
+
+    }
+
+    /**
+     * The current thread wait until all the dependencies are processed
+     * @throws ExecutionException
+     */
+    public void waitUntilResolveAllDependencies() throws ExecutionException {
+        final String submitterName = "DependencyManagerSubmitter" + Thread.currentThread().getName();
+        submitter = DotConcurrentFactory.getInstance().getSubmitter(submitterName,
+                new DotConcurrentFactory.SubmitterConfigBuilder()
+                        .poolSize(
+                                Config.getIntProperty("MIN_NUMBER_THREAD_TO_EXECUTE_BUNDLER", 10))
+                        .maxPoolSize(Config.getIntProperty("MAX_NUMBER_THREAD_TO_EXECUTE_BUNDLER", 40))
+                        .queueCapacity(Config.getIntProperty("QUEUE_CAPACITY_TO_EXECUTE_BUNDLER", Integer.MAX_VALUE))
+                        .build()
+        );
+
+        try {
+            while (!isFinish()) {
+                try {
+                    Logger.debug(ConcurrentDependencyProcessor.class, () -> "Waiting for more assets");
+                    final DependencyProcessorItem dependencyProcessorItem = queue.take();
+                    Logger.debug(ConcurrentDependencyProcessor.class,
+                            () -> "Taking one " + dependencyProcessorItem.asset);
+                    if (!dependencyProcessorItem
+                            .equals(DependencyProcessorItem.FINISHED_DEPENDENCY_PROCESSOR_ITEM)) {
+                        submitter.submit(new DependencyRunnable(dependencyProcessorItem));
+                    } else {
+                        finishReceived.incrementAndGet();
+                    }
+                } catch (InterruptedException e) {
+                    throw new RuntimeException(e);
+                }
+            }
+            Logger.info(ConcurrentDependencyProcessor.class, "DependencyProcessor Finished");
+        } finally {
+            submitter.shutdownNow();
+        }
+    }
+
+    private synchronized void sendFinishNotification() {
+        queue.add(DependencyProcessorItem.FINISHED_DEPENDENCY_PROCESSOR_ITEM);
+    }
+
+    private boolean isFinish() {
+        return queue.isEmpty() && isAllTaskFinished();
+    }
+
+    private synchronized boolean isAllTaskFinished() {
```

Comment:
same here, if all collections and objects are thread-safe probably do not need sync method

Reply:
we already do a lot of test with this synchronized so at least for now I think is better keep it in this way